atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java

```diff
@@ -209,7 +211,7 @@ Iterator<Map.Entry<Cell, byte[]>> getSortedColumns(
             allowedOnPath = ".*/src/test/.*", // Unsafe behavior in tests is ok.
             allowlistAnnotations = {ReviewedRestrictedApiUsage.class})
     @Idempotent
-    Map<Cell, byte[]> getWithExpectedNumberOfCells(
+    Result<Map<Cell, byte[]>, MoreCellsPresentThanExpectedException> getWithExpectedNumberOfCells(
```

Comment:
I'm not opposed to introducing this type/results of this manner to this class. The only thing to consider here is what types of errors we return.  I see in this case it would be explicitly for `MoreCellsPresentThanExpectedException`, but from a user perspective it may seem odd to me that I'm given the chance to handle some exceptions but not others. But maybe that's ok, as arguably the other exceptions  are retryable or not of importance?

Only other thing we need to me mindful here is that, when this exception is caught and wrapped, we still need to ensure we abort the transaction -- but I guess we are sort of delegating that to the caller to make that determination?

Reply:
In the way Result types are usually applied, you're right that in theory we should handle all errors through it, and that's not what we're doing here.

The problem here came from a perspective that we wanted a behaviour like checked exceptions for this method: forcing the user to explicitly handle this bad case. But that would not be feasible right now without breaking the API of all the other get methods that already exist, which would need to declare the new check exception and force all clients to handle them. So we made the conscious choice of forcing only on `getWithExpectedNumberOfCells` users to explicitly handle the possibility of a `MoreCellsPresentThanExpectedException` by the use of Result types. Other exceptions can still exist, but don't need to be treated any different from the ways they're treated in our already existing get methods.

Reply:
Regarding aborting the transaction: yes we're delegating it to the caller to make the determination. But you might have a point that we should abort on our side? As in, when that error occurs it means there there is a bug in the assumptions of how the client is using the API (or KV corruption) and not much can be relied on after that fact